akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +351,120 @@ final class CoordinatedShutdown private[akka] (
     system: ExtendedActorSystem,
     phases: Map[String, CoordinatedShutdown.Phase])
     extends Extension {
-  import CoordinatedShutdown.Reason
-  import CoordinatedShutdown.UnknownReason
+  import CoordinatedShutdown.{ Reason, UnknownReason }
 
   /** INTERNAL API */
   private[akka] val log = Logging(system, getClass)
   private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
 
   /** INTERNAL API */
   private[akka] val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
-  private val tasks = new ConcurrentHashMap[String, Vector[(String, () => Future[Done])]]
```

Comment:
Rather big refactoring here. We could compare with the alternative of keeping the original ConcurrentHashMap structure and implement it like this:

```scala
@tailrec def addCancellableTask(phase: String, taskName: String)(task: () => Future[Done]): Cancellable = {
    require(
      knownPhases(phase),
      s"Unknown phase [$phase], known phases [$knownPhases]. " +
      "All phases (along with their optional dependencies) must be defined in configuration")
    require(
      taskName.nonEmpty,
      "Set a task name when adding tasks to the Coordinated Shutdown. " +
      "Try to use unique, self-explanatory names.")

    val taskEntry = taskName -> task
    val cancellable = new Cancellable {
      override def cancel(): Boolean =
        removeTask(phase, taskEntry)

      override def isCancelled: Boolean =
        !tasks.get(phase).exists(_ eq taskEntry)
    }

    val current = tasks.get(phase)
    if (current == null) {
      if (tasks.putIfAbsent(phase, Vector(taskEntry)) == null)
        cancellable
      else
        addCancellableTask(phase, taskName)(task) // CAS failed, retry
    } else {
      if (tasks.replace(phase, current, current :+ taskEntry))
        cancellable
      else
        addCancellableTask(phase, taskName)(task) // CAS failed, retry
    }
  }

  @tailrec private def removeTask(phase: String, entry: (String, () => Future[Done])): Boolean = {
    val currentTasks = tasks.get(phase)
    val newTasks = currentTasks.filterNot(_ eq entry)
    if (newTasks.size == currentTasks.size)
      false // already removed
    else if (tasks.replace(phase, currentTasks, newTasks))
      true // removed
    else
      removeTask(phase, entry) // CAS failed, retry
  }
```

Reply:
This is actually one of the reasons I wrote a larger refactor. The solution above uses low-level methods of `ConcurrentHashMap` (one-shot thin wrappers over `compareAndSet`) instead of ones at the level of abstraction of the rest of the code here. I personally think that a lot of the public API of various classes in `java.util.concurrent` should be private or protected, and that it's at best a code smell to perform side-effecting operations in a condition.

I do understand the value in replacing a naive-recursive method call with a tail-recursive alternative, but that's not the case here (as it is with some of the higher-level methods on `AtomicWhatever`).